eth/consensus/clique/_utils.py

```diff
@@ -0,0 +1,154 @@
+import time
+from typing import (
+    Iterable,
+)
+
+from eth.abc import (
+    BlockHeaderAPI,
+)
+from eth.constants import (
+    EMPTY_UNCLE_HASH,
+    ZERO_ADDRESS,
+    ZERO_HASH32,
+)
+from eth_keys import keys
+from eth_keys.datatypes import PrivateKey
+from eth_typing import (
+    Address,
+    Hash32,
+)
+from eth_utils import (
+    encode_hex,
+    to_list,
+    ValidationError,
+)
+
+from .constants import (
+    COMMON_ADDRESS_LENGTH,
+    NONCE_AUTH,
+    NONCE_DROP,
+    SIGNATURE_LENGTH,
+    VANITY_LENGTH,
+)
+from .datatypes import (
+    Snapshot,
+)
+
+
+@to_list
+def get_signers_at_checkpoint(header: BlockHeaderAPI) -> Iterable[Address]:
+    """
+    Read the list of signers from a checkpoint header.
+    """
+
+    signer_count = int(
+        (len(header.extra_data) - VANITY_LENGTH - SIGNATURE_LENGTH) / COMMON_ADDRESS_LENGTH
+    )
+
+    for i in range(signer_count):
+        yield Address(
+            header.extra_data[VANITY_LENGTH + i * COMMON_ADDRESS_LENGTH:][:COMMON_ADDRESS_LENGTH]
+        )
+
+
+def get_signature_hash(header: BlockHeaderAPI) -> Hash32:
+    """
+    Return the hash that is signed by the block producer. It is defined as the hash of
+    the ``header`` except that the last 65 bytes of the ``extra_data`` (the signature)
+    are removed before calculating the hash.
+    """
+    signature_header = header.copy(
+        extra_data=header.extra_data[:len(header.extra_data) - SIGNATURE_LENGTH]
+    )
+    return signature_header.hash
+
+
+def get_block_signer(header: BlockHeaderAPI) -> Address:
+    """
+    Return the address of the signer of the ``header``.
+    """
+
+    if len(header.extra_data) < SIGNATURE_LENGTH:
+        raise Exception("header.extra_data too short to contain signature")
+
+    signature_bytes = header.extra_data[-SIGNATURE_LENGTH:]
+
+    signature = keys.Signature(signature_bytes=signature_bytes)
+
+    signature_hash = get_signature_hash(header)
+    return signature.recover_public_key_from_msg_hash(signature_hash).to_canonical_address()
+
+
+def is_in_turn(signer: Address, snapshot: Snapshot, header: BlockHeaderAPI) -> bool:
+    """
+    Return ``True`` if the block was produced *in turn*, otherwise return ``False``.
+    """
+    sorted_signers = snapshot.get_sorted_signers()
+    offset = 0
+    while offset < len(sorted_signers) and sorted_signers[offset] != signer:
+        offset += 1
+
+    return header.block_number % len(sorted_signers) == offset
+
+
+def sign_block_header(header: BlockHeaderAPI, private_key: PrivateKey) -> BlockHeaderAPI:
+    signature_hash = get_signature_hash(header)
+    signature = private_key.sign_msg_hash(signature_hash)
+
+    signers = get_signers_at_checkpoint(header)
+    signed_extra_data = header.extra_data[:VANITY_LENGTH]
+    for signer in signers:
+        signed_extra_data += signer
+
+    signed_extra_data += signature.to_bytes()
```

Comment:
I don't know if it actually saves the object copies, but you can remove the mutability here with:

```python
signed_extra_data = b''.join(
    header.extra_data[...],
    b''.join(signers),
    signature.to_bytes(),
)
```

Alternatively you could use a `bytearray` which is inherently mutable and then cast it to `bytes` once's you've built up the state.

Reply:
Looks good :+1: